modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/CurrencySnapshotCreator.scala

```diff
@@ -101,29 +101,36 @@ object CurrencySnapshotCreator {
               .map(_.toOption)
               .handleErrorWith(err =>
                 logger.error(err)(s"Unhandled exception during deserialization data application, fallback to empty state") >>
-                  none[DataState].pure[F]
+                  none[DataOnChainState].pure[F]
               )
         }.flatSequence
 
-        maybeNewDataState <- (maybeDataApplicationService, maybeLastDataState).mapN {
-          case ((nodeContext, service), lastState) =>
+        maybeNewDataState <- (maybeDataApplicationService, maybeLastDataOnChainState).mapN {
+          case (((nodeContext, service), lastState)) =>
             implicit val context: L0NodeContext[F] = nodeContext
             NonEmptyList
               .fromList(dataBlocks.distinctBy(_.value.roundId))
               .map(_.flatMap(_.value.updates))
               .map { updates =>
-                service
-                  .validateData(lastState, updates)
-                  .flatTap { validated =>
-                    Applicative[F].unlessA(validated.isValid)(logger.warn(s"Data application is invalid, errors: ${validated.toString}"))
-                  }
+                service.getCalculatedState.flatMap { off =>
+                  service.validateData(DataState(lastState, off), updates)
+                }.flatTap { validated =>
+                  Applicative[F].unlessA(validated.isValid)(logger.warn(s"Data application is invalid, errors: ${validated.toString}"))
+                }
                   .map(_.isValid)
-                  .handleErrorWith(logger.error(_)(s"Unhandled exception during validating data application, assumed as invalid").as(false))
+                  .handleErrorWith(err =>
+                    logger.error(err)(s"Unhandled exception during validating data application, assumed as invalid").as(false)
+                  )
                   .ifF(updates.toList, List.empty[Signed[DataUpdate]])
               }
               .getOrElse(List.empty[Signed[DataUpdate]].pure[F])
-              .flatMap(service.combine(lastState, _))
-              .flatMap(service.serializeState)
+              .flatMap { updates =>
+                service.getCalculatedState
+                  .map(DataState(lastState, _))
+                  .flatMap(service.combine(_, updates))
+                  .flatTap(state => service.setCalculatedState(state.calculated))
```

Comment:
Should we set state in the middle of the function run? I mean what if the function fails? The state will be set. Shouldn't we return the off chain state as the output of this function?

Reply:
Let me revisit this later on. I agree with you - in case of partial failure, we may end up with corrupted state. Anyway, we're safe because of other checks (ordinal and hash proof), so nodes should not progress with corrupted state.